### ISSUE TITLE
cannot resolve stream adapter for kotlin coroutines

### ISSUE THREAD

**Issue by pooyakn** [Tinder/Scarlet#54]
Hi,
The same code using RxJava stream adapter works fine.

```kotlin
val scarlet = Scarlet.Builder().webSocketFactory(client.newWebSocketFactory(SOCKET_URL))
            .lifecycle(lifecycle)
            .addMessageAdapterFactory(GsonMessageAdapter.Factory())
            .addStreamAdapterFactory(CoroutinesStreamAdapterFactory()).build()
return scarlet.create()
```

```kotlin
interface CashCoolSocketService {
    @Receive
    fun observeUser(): ReceiveChannel<UserSerializer>
}
```

```kotlin
GlobalScope.launch(CoroutineExceptionHandler { _, throwable ->
            Log.e(
                "SOCKET ERROR",
                throwable.message
            )
        }) {
            val user = ws.observeUser().receive()
            Log.d("SOCKET MESSAGE", user.toString())
        }
```
<details>
<summary>Throws following error:</summary>
<code>
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: app.cashcool.cashcool, PID: 3109
    java.lang.IllegalStateException: Cannot resolve stream adapter for type kotlinx.coroutines.channels.ReceiveChannel<app.cashcool.cashcool.features.user.UserSerializer>.
        at com.tinder.scarlet.internal.servicemethod.StreamAdapterResolver.resolve(StreamAdapterResolver.kt:26)
        at com.tinder.scarlet.internal.servicemethod.ServiceMethod$Receive$Factory.createStreamAdapter(ServiceMethod.kt:89)
        at com.tinder.scarlet.internal.servicemethod.ServiceMethod$Receive$Factory.create(ServiceMethod.kt:81)
        at com.tinder.scarlet.internal.servicemethod.ServiceMethod$Receive$Factory.create(ServiceMethod.kt:66)
        at com.tinder.scarlet.internal.servicemethod.ServiceMethodExecutor$Factory.toServiceMethod(ServiceMethodExecutor.kt:47)
        at com.tinder.scarlet.internal.servicemethod.ServiceMethodExecutor$Factory.findServiceMethods(ServiceMethodExecutor.kt:38)
        at com.tinder.scarlet.internal.servicemethod.ServiceMethodExecutor$Factory.create(ServiceMethodExecutor.kt:32)
        at com.tinder.scarlet.internal.Service$Factory.create(Service.kt:28)
        at com.tinder.scarlet.Scarlet.implementService(Scarlet.kt:108)
        at com.tinder.scarlet.Scarlet.create(Scarlet.kt:100)
        at app.cashcool.cashcool.core.di.ApplicationModule.provideCashCoolSocketService(ApplicationModule.kt:109)
        at app.cashcool.cashcool.core.di.ApplicationModule_ProvideCashCoolSocketServiceFactory.proxyProvideCashCoolSocketService(ApplicationModule_ProvideCashCoolSocketServiceFactory.java:55)
        at app.cashcool.cashcool.core.di.ApplicationModule_ProvideCashCoolSocketServiceFactory.provideInstance(ApplicationModule_ProvideCashCoolSocketServiceFactory.java:41)
        at app.cashcool.cashcool.core.di.ApplicationModule_ProvideCashCoolSocketServiceFactory.get(ApplicationModule_ProvideCashCoolSocketServiceFactory.java:34)
        at app.cashcool.cashcool.core.di.ApplicationModule_ProvideCashCoolSocketServiceFactory.get(ApplicationModule_ProvideCashCoolSocketServiceFactory.java:11)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at app.cashcool.cashcool.core.di.DaggerApplicationComponent.injectCreateUserFragment(DaggerApplicationComponent.java:328)
        at app.cashcool.cashcool.core.di.DaggerApplicationComponent.inject(DaggerApplicationComponent.java:288)
        at app.cashcool.cashcool.features.user.create.CreateUserFragment.onCreate(CreateUserFragment.kt:109)
        at androidx.fragment.app.Fragment.performCreate(Fragment.java:2414)
        at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManagerImpl.java:844)
        at androidx.fragment.app.FragmentTransition.addToFirstInLastOut(FragmentTransition.java:1196)
        at androidx.fragment.app.FragmentTransition.calculateFragments(FragmentTransition.java:1079)
        at androidx.fragment.app.FragmentTransition.startTransitions(FragmentTransition.java:118)
        at androidx.fragment.app.FragmentManagerImpl.executeOpsTogether(FragmentManagerImpl.java:1844)
        at androidx.fragment.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManagerImpl.java:1802)
        at androidx.fragment.app.FragmentManagerImpl.execPendingActions(FragmentManagerImpl.java:1709)
        at androidx.fragment.app.FragmentManagerImpl$1.run(FragmentManagerImpl.java:147)
        at android.os.Handler.handleCallback(Handler.java:789)
        at android.os.Handler.dispatchMessage(Handler.java:98)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6541)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
     Caused by: io.reactivex.exceptions.CompositeException: 2 exceptions occurred. 
        at com.tinder.scarlet.internal.servicemethod.StreamAdapterResolver.resolve(StreamAdapterResolver.kt:25)
        at com.tinder.scarlet.internal.servicemethod.ServiceMethod$Receive$Factory.createStreamAdapter(ServiceMethod.kt:89) 
        at com.tinder.scarlet.internal.servicemethod.ServiceMethod$Receive$Factory.create(ServiceMethod.kt:81) 
        at com.tinder.scarlet.internal.servicemethod.ServiceMethod$Receive$Factory.create(ServiceMethod.kt:66) 
        at com.tinder.scarlet.internal.servicemethod.ServiceMethodExecutor$Factory.toServiceMethod(ServiceMethodExecutor.kt:47) 
        at com.tinder.scarlet.internal.servicemethod.ServiceMethodExecutor$Factory.findServiceMethods(ServiceMethodExecutor.kt:38) 
        at com.tinder.scarlet.internal.servicemethod.ServiceMethodExecutor$Factory.create(ServiceMethodExecutor.kt:32) 
        at com.tinder.scarlet.internal.Service$Factory.create(Service.kt:28) 
        at com.tinder.scarlet.Scarlet.implementService(Scarlet.kt:108) 
        at com.tinder.scarlet.Scarlet.create(Scarlet.kt:100) 
        at app.cashcool.cashcool.core.di.ApplicationModule.provideCashCoolSocketService(ApplicationModule.kt:109) 
        at app.cashcool.cashcool.core.di.ApplicationModule_ProvideCashCoolSocketServiceFactory.proxyProvideCashCoolSocketService(ApplicationModule_ProvideCashCoolSocketServiceFactory.java:55) 
        at app.cashcool.cashcool.core.di.ApplicationModule_ProvideCashCoolSocketServiceFactory.provideInstance(ApplicationModule_ProvideCashCoolSocketServiceFactory.java:41) 
        at app.cashcool.cashcool.core.di.ApplicationModule_ProvideCashCoolSocketServiceFactory.get(ApplicationModule_ProvideCashCoolSocketServiceFactory.java:34) 
        at app.cashcool.cashcool.core.di.ApplicationModule_ProvideCashCoolSocketServiceFactory.get(ApplicationModule_ProvideCashCoolSocketServiceFactory.java:11) 
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47) 
        at app.cashcool.cashcool.core.di.DaggerApplicationComponent.injectCreateUserFragment(DaggerApplicationComponent.java:328) 
        at app.cashcool.cashcool.core.di.DaggerApplicationComponent.inject(DaggerApplicationComponent.java:288) 
        at app.cashcool.cashcool.features.user.create.CreateUserFragment.onCreate(CreateUserFragment.kt:109) 
        at androidx.fragment.app.Fragment.performCreate(Fragment.java:2414) 
        at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManagerImpl.java:844) 
        at androidx.fragment.app.FragmentTransition.addToFirstInLastOut(FragmentTransition.java:1196) 
        at androidx.fragment.app.FragmentTransition.calculateFragments(FragmentTransition.java:1079) 
        at androidx.fragment.app.FragmentTransition.startTransitions(FragmentTransition.java:118) 
        at androidx.fragment.app.FragmentManagerImpl.executeOpsTogether(FragmentManagerImpl.java:1844) 
        at androidx.fragment.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManagerImpl.java:1802) 
        at androidx.fragment.app.FragmentManagerImpl.execPendingActions(FragmentManagerImpl.java:1709) 
        at androidx.fragment.app.FragmentManagerImpl$1.run(FragmentManagerImpl.java:147) 
        at android.os.Handler.handleCallback(Handler.java:789) 
        at android.os.Handler.dispatchMessage(Handler.java:98) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6541) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767) 
     Caused by: io.reactivex.exceptions.CompositeException$CompositeExceptionCausalChain: Chain of Causes for CompositeException In Order Received =>
        at android.util.Log.printlns(Log.java:417)
E/AndroidRuntime:     at com.android.internal.os.RuntimeInit.Clog_e(RuntimeInit.java:60)
        at com.android.internal.os.RuntimeInit.-wrap0(Unknown Source:0)
        at com.android.internal.os.RuntimeInit$LoggingHandler.uncaughtException(RuntimeInit.java:89)
        at java.lang.Thread.dispatchUncaughtException(Thread.java:1948)
     Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Lkotlinx/coroutines/experimental/channels/ReceiveChannel;
        at com.tinder.streamadapter.coroutines.CoroutinesStreamAdapterFactory.create(CoroutinesStreamAdapterFactory.kt:19)
        at com.tinder.scarlet.internal.servicemethod.StreamAdapterResolver.resolve(StreamAdapterResolver.kt:19)
        at com.tinder.scarlet.internal.servicemethod.ServiceMethod$Receive$Factory.createStreamAdapter(ServiceMethod.kt:89)
        at com.tinder.scarlet.internal.servicemethod.ServiceMethod$Receive$Factory.create(ServiceMethod.kt:81)
        at com.tinder.scarlet.internal.servicemethod.ServiceMethod$Receive$Factory.create(ServiceMethod.kt:66)
        at com.tinder.scarlet.internal.servicemethod.ServiceMethodExecutor$Factory.toServiceMethod(ServiceMethodExecutor.kt:47)
        at com.tinder.scarlet.internal.servicemethod.ServiceMethodExecutor$Factory.findServiceMethods(ServiceMethodExecutor.kt:38)
        at com.tinder.scarlet.internal.servicemethod.ServiceMethodExecutor$Factory.create(ServiceMethodExecutor.kt:32)
        at com.tinder.scarlet.internal.Service$Factory.create(Service.kt:28)
        at com.tinder.scarlet.Scarlet.implementService(Scarlet.kt:108)
        at com.tinder.scarlet.Scarlet.create(Scarlet.kt:100)
        at app.cashcool.cashcool.core.di.ApplicationModule.provideCashCoolSocketService(ApplicationModule.kt:109)
        at app.cashcool.cashcool.core.di.ApplicationModule_ProvideCashCoolSocketServiceFactory.proxyProvideCashCoolSocketService(ApplicationModule_ProvideCashCoolSocketServiceFactory.java:55)
        at app.cashcool.cashcool.core.di.ApplicationModule_ProvideCashCoolSocketServiceFactory.provideInstance(ApplicationModule_ProvideCashCoolSocketServiceFactory.java:41)
        at app.cashcool.cashcool.core.di.ApplicationModule_ProvideCashCoolSocketServiceFactory.get(ApplicationModule_ProvideCashCoolSocketServiceFactory.java:34)
        at app.cashcool.cashcool.core.di.ApplicationModule_ProvideCashCoolSocketServiceFactory.get(ApplicationModule_ProvideCashCoolSocketServiceFactory.java:11)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at app.cashcool.cashcool.core.di.DaggerApplicationComponent.injectCreateUserFragment(DaggerApplicationComponent.java:328)
        at app.cashcool.cashcool.core.di.DaggerApplicationComponent.inject(DaggerApplicationComponent.java:288)
        at app.cashcool.cashcool.features.user.create.CreateUserFragment.onCreate(CreateUserFragment.kt:109)
        at androidx.fragment.app.Fragment.performCreate(Fragment.java:2414)
        at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManagerImpl.java:844)
        at androidx.fragment.app.FragmentTransition.addToFirstInLastOut(FragmentTransition.java:1196)
        at androidx.fragment.app.FragmentTransition.calculateFragments(FragmentTransition.java:1079)
        at androidx.fragment.app.FragmentTransition.startTransitions(FragmentTransition.java:118)
        at androidx.fragment.app.FragmentManagerImpl.executeOpsTogether(FragmentManagerImpl.java:1844)
        at androidx.fragment.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManagerImpl.java:1802)
        at androidx.fragment.app.FragmentManagerImpl.execPendingActions(FragmentManagerImpl.java:1709)
        at androidx.fragment.app.FragmentManagerImpl$1.run(FragmentManagerImpl.java:147)
        at android.os.Handler.handleCallback(Handler.java:789)
        at android.os.Handler.dispatchMessage(Handler.java:98)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6541)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
E/AndroidRuntime: Caused by: java.lang.ClassNotFoundException: Didn't find class "kotlinx.coroutines.experimental.channels.ReceiveChannel" on path: DexPathList[[zip file "/data/app/app.cashcool.cashcool-Szw56xvjCrKy9lGGtl4lKg==/base.apk", zip file "/data/app/app.cashcool.cashcool-Szw56xvjCrKy9lGGtl4lKg==/split_lib_dependencies_apk.apk", zip file "/data/app/app.cashcool.cashcool-Szw56xvjCrKy9lGGtl4lKg==/split_lib_resources_apk.apk", zip file "/data/app/app.cashcool.cashcool-Szw56xvjCrKy9lGGtl4lKg==/split_lib_slice_0_apk.apk", zip file "/data/app/app.cashcool.cashcool-Szw56xvjCrKy9lGGtl4lKg==/split_lib_slice_1_apk.apk", zip file "/data/app/app.cashcool.cashcool-Szw56xvjCrKy9lGGtl4lKg==/split_lib_slice_2_apk.apk", zip file "/data/app/app.cashcool.cashcool-Szw56xvjCrKy9lGGtl4lKg==/split_lib_slice_3_apk.apk", zip file "/data/app/app.cashcool.cashcool-Szw56xvjCrKy9lGGtl4lKg==/split_lib_slice_4_apk.apk", zip file "/data/app/app.cashcool.cashcool-Szw56xvjCrKy9lGGtl4lKg==/split_lib_slice_5_apk.apk", zip file "/data/app/app.cashcool.cashcool-Szw56xvjCrKy9lGGtl4lKg==/split_lib_slice_6_apk.apk", zip file "/data/app/app.cashcool.cashcool-Szw56xvjCrKy9lGGtl4lKg==/split_lib_slice_7_apk.apk", zip file "/data/app/app.cashcool.cashcool-Szw56xvjCrKy9lGGtl4lKg==/split_lib_slice_8_apk.apk", zip file "/data/app/app.cashcool.cashcool-Szw56xvjCrKy9lGGtl4lKg==/split_lib_slice_9_apk.apk"],nativeLibraryDirectories=[/data/app/app.cashcool.cashcool-Szw56xvjCrKy9lGGtl4lKg==/lib/x86, /system/lib, /system/vendor/lib]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:93)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        	... 36 more
    	Suppressed: java.io.IOException: No original dex files found for dex location /data/app/app.cashcool.cashcool-Szw56xvjCrKy9lGGtl4lKg==/split_lib_resources_apk.apk
        at dalvik.system.DexFile.openDexFileNative(Native Method)
        at dalvik.system.DexFile.openDexFile(DexFile.java:353)
        at dalvik.system.DexFile.<init>(DexFile.java:100)
        at dalvik.system.DexFile.<init>(DexFile.java:74)
        at dalvik.system.DexPathList.loadDexFile(DexPathList.java:374)
        at dalvik.system.DexPathList.makeDexElements(DexPathList.java:337)
        at dalvik.system.DexPathList.<init>(DexPathList.java:157)
        at dalvik.system.BaseDexClassLoader.<init>(BaseDexClassLoader.java:65)
        at dalvik.system.PathClassLoader.<init>(PathClassLoader.java:64)
        at com.android.internal.os.PathClassLoaderFactory.createClassLoader(PathClassLoaderFactory.java:43)
        at android.app.ApplicationLoaders.getClassLoader(ApplicationLoaders.java:69)
        at android.app.ApplicationLoaders.getClassLoader(ApplicationLoaders.java:36)
        at android.app.LoadedApk.createOrUpdateClassLoaderLocked(LoadedApk.java:676)
        at android.app.LoadedApk.getClassLoader(LoadedApk.java:709)
        at android.app.LoadedApk.getResources(LoadedApk.java:936)
        at android.app.ContextImpl.createAppContext(ContextImpl.java:2242)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5672)
        at android.app.ActivityThread.-wrap1(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1661)
        at android.os.Handler.dispatchMessage(Handler.java:105)
        		... 5 more
<code>
</details>

**Comment by zhxnlai** [comment]
@pooyakn Looks like your stacktrace contains

```
java.lang.NoClassDefFoundError: Failed resolution of: Lkotlinx/coroutines/experimental/channels/ReceiveChannel
```

This is because the `CoroutinesStreamAdapterFactory` in `0.2.1-alpha4` uses an experimental version of kotlin coroutines. I bumped the coroutines version and published `0.2.1-alpha5`. Let me know if `0.2.1-alpha5` solves this issue.

**Comment by pooyakn** [comment]
Thanks. It runs now with 0.2.1-alpha5.
